tests/ga/test_exthandlers_exthandlerinstance.py

```diff
@@ -0,0 +1,95 @@
+
+    def test_rm_ext_handler_dir_should_report_an_event_if_an_error_occurrs_while_deleting_the_extension_directory(self):
+        extension_directory = os.path.join(self.tmp_dir, "extension_directory")
+        os.mkdir(extension_directory)
+        os.mknod(os.path.join(extension_directory, "extension_file1"))
+        os.mknod(os.path.join(extension_directory, "extension_file2"))
+        os.mknod(os.path.join(extension_directory, "extension_file3"))
+
+        original_remove = shutil.os.remove
+
+        def mock_remove(path):
+            if path.endswith("extension_file2"):
+                mock_remove.exception_raised = True
+                raise IOError("A mocked error")
+            original_remove(path)
+
+        with patch.object(shutil.os, "remove", mock_remove) as mock_remove:
+            with patch.object(self.ext_handler_instance, "report_event") as mock_report_event:
+                self.ext_handler_instance.rm_ext_handler_dir()
+
+        args, kwargs = mock_report_event.call_args
+        self.assertTrue("A mocked error" in kwargs["message"])
+
+        self.assertTrue(mock_remove.exception_raised)  # verify the mock was actually called
+
+    def test_rm_ext_handler_dir_should_not_report_an_event_if_the_extension_directory_does_not_exist(self):
+        if os.path.exists(self.extension_directory):
+            os.rmdir(self.extension_directory)
+
+        with patch.object(self.ext_handler_instance, "report_event") as mock_report_event:
+            self.ext_handler_instance.rm_ext_handler_dir()
+
+        mock_report_event.assert_not_called()
+
+    def test_rm_ext_handler_dir_should_not_report_an_event_if_a_child_is_removed_asynchronously(self):
+        os.mkdir(self.extension_directory)
+        os.mknod(os.path.join(self.extension_directory, "extension_file1"))
+        os.mknod(os.path.join(self.extension_directory, "extension_file2"))
+        os.mknod(os.path.join(self.extension_directory, "extension_file3"))
+
+        # Some extensions uninstall asynchronously and the files we are trying to remove may be removed
+        # while shutil.rmtree is traversing the extension's directory. Mock this by deleting a file
+        # twice (the second call will produce "[Errno 2] No such file or directory", which should not be
+        # reported as a telemetry event.
+        original_remove = shutil.os.remove
+
+        def mock_remove(path):
+            if path.endswith("extension_file2"):
+                original_remove(path)
+                mock_remove.file_deleted_asynchronously = True
+            original_remove(path)
+
+        with patch.object(shutil.os, "remove", mock_remove) as mock_remove:
+            with patch.object(self.ext_handler_instance, "report_event") as mock_report_event:
+                self.ext_handler_instance.rm_ext_handler_dir()
+
+        mock_report_event.assert_not_called()
+
+        self.assertTrue(mock_remove.file_deleted_asynchronously)  # verify the mock was actually called
```

Comment:
This is also failing in Python 3.x, I think the mocks are not getting called for some reason